posts/2018-07-12-servant-dsl-typelevel.md

```diff
@@ -0,0 +1,664 @@
+
+``` haskell
+infixr 5 :>
+(:>) :: (Endpoint -> Endpoint) -> Endpoint -> Endpoint
+f :> x = f x
+
+getHelloNew :: Endpoint
+getHelloNew = Static "hello" :> Capture :> Verb Get
+```
+
+Unlike servant though, as you can see with the type of `getHello` and
+`getHelloNew`, our descriptions are good old Haskell values, both of the
+`Endpoint` type.
+
+Given those few definitions, how could we go about, say, generating links
+to endpoints? Well, here is a straightforward attempt.
+
+``` haskell
+-- a link here is just a list of path components
+-- (we ignore query parameters in this post)
+type Link = [String]
+
+linkTo :: Endpoint -> Link
+linkTo (Static str rest) = str : linkTo rest
+linkTo (Verb _method)    = []
+linkTo (Capture rest)    = ??? : linkTo rest
+```
+
+But... what should we put in place of those `???`, if anything?
+
+Well, we definitely want to add _some_ path component, to fill the
+`Capture` slot. However, by definition, a captured path fragment is
+not fixed, it is allowed to vary. In other words, `Capture :> Verb Post`
+matches both `POST /x` and `POST /y`. We cannot just pick one value
+and hope that it is the one the user wanted. We need to take it as an
+argument. But what about `Capture :> Capture :> Verb Post`? We would need
+our `linkTo` function to take 2 arguments for that case. And zero additional
+argument for `Static "hello" :> Verb Post`. This is quite problematic.
+
+Indeed, we would like the type of `linkTo` to be `Endpoint -> Link`,
+`Endpoint -> String -> Link`, `Endpoint -> String -> String -> Link` and
+so on depending on what the `Endpoint` argument is. In other words,
+we want the return type of `linkTo` (when really seen as a function of
+one argument, which it is anyway) to depend on the value of type `Endpoint`
+it gets as input. That is, we want a type that depends on a value, i.e
+dependent types.
+
+The alternative would be:
+
+``` haskell
+-- linkTo assumes that the capture values are given in the same order as we want
+-- them to appear in the url.
+linkTo :: Endpoint -> [String] -> Link
+linkTo (Static str rest) captureValues  = str : linkTo rest captureValues
+-- when there is at least one value left in the list, we use it.
+-- when there isn't... we error out.
+linkTo (Capture rest)    (c : cs)       = c   : linkTo rest cs
+linkTo (Capture rest)    []             = error "linkTo: capture value needed but the list is empty" -- :-(
+linkTo (Verb method)     _              = []
+```
+
+This solution is very unsatisfactory, first and foremost because it is possible
+for it to error out if we don't supply the right number of captures. However,
+it will also silently let us pass too many capture values without telling us
+that some of them are not used. Such a function should be total, we really don't
+want an implementation that can let us down if we're not very careful.
+
+Fortunately, GADTs can help here. We could turn `Endpoint` into
```

Comment:
can we squeeze the comment that GADT approach is good  (and maybe more approachable) in some very stable domains (and HTTP API design isn't).

Reply:
done, a bit below, in an upcoming commit